### PR TITLE
Non type class absorbers

### PR DESCRIPTION
**Pull request by adamConnerSax**
This is what I have in mind.  The code for each absorber is simpler and the typeclass stuff is all gone, since I think all of its cleverness was only useful for the absorbers in the MTL module.  This way we are setting a good example in the zoo by doing each one separately and exporting only the constraint-monomorphic absorber. And we are setting a good naming example for people to follow if they write their own absorbers for others to use.

Tests all pass unchanged.  The only change there was the name of the test file and the imports.  Which is reassuring.

That being said, it's a breaking change.

I'm going to re-write RandomFu's absorber this way and my PandocMonad absorber just to make sure it all works as simply as before, etc.

### CODE REVIEW

**Comment by adamConnerSax** [comment]
Also, I'm not sure why it has so many commits.  I thought I had merged the upstream stuff before doing this work.  Is that an issue?  Is there something I can do on my end to fix it?

**Comment by adamConnerSax** [comment]
One more question:  We don't need to export "using" from Polysemy.ConstraintAbsorber.  Should I drop it since ```absorbWithSem``` is the version we want people to use since it has the right signature and simpler order of type-variables for use in making absorbers?

File: src/Polysemy/ConstraintAbsorber.hs
Comment (isovector) [inline src/Polysemy/ConstraintAbsorber.hs]: This is already specified in the cabal file, so unnecessary here.

File: src/Polysemy/ConstraintAbsorber.hs
Comment (isovector) [inline src/Polysemy/ConstraintAbsorber.hs]: STV and TypeFamilies are already enabled on a project-wide basis. Please remove from here and other files.

File: src/Polysemy/ConstraintAbsorber.hs
Comment (isovector) [inline src/Polysemy/ConstraintAbsorber.hs]: Any reason we can't just inline the definition of `using` here, and then drop `using` entirely?

File: src/Polysemy/ConstraintAbsorber.hs
Comment (isovector) [inline src/Polysemy/ConstraintAbsorber.hs]: Nit: please style types like this:

```haskell
foo
    :: forall (a :: k1)
              (b :: k2)
     . Monad m
    => d
```

File: src/Polysemy/ConstraintAbsorber/MonadError.hs
Comment (isovector) [inline src/Polysemy/ConstraintAbsorber/MonadError.hs]: It might be clearer to name this `Orphan`.

File: src/Polysemy/ConstraintAbsorber/MonadError.hs
Comment (isovector) [inline src/Polysemy/ConstraintAbsorber/MonadError.hs]: Nit: here and elsewhere, prefer to put the first element on the same line as the opener, eg:

```haskell
data Foo = Foo
  { throwError_ :: ..
```

likewise for module export lists

File: src/Polysemy/ConstraintAbsorber/MonadError.hs
Comment (isovector) [inline src/Polysemy/ConstraintAbsorber/MonadError.hs]: I just realized I have no idea how this works. How come giving an instance for `Action` gives us an instance for `Sem`? Do we unsafe coerce the dict somewhere? Is this how it used to work too, and I've just been cargo-culting all along?

File: src/Polysemy/ConstraintAbsorber/MonadWriter.hs
Comment (isovector) [inline src/Polysemy/ConstraintAbsorber/MonadWriter.hs]: Is it possible to factor this `Sub Dict` into `absorbWithSem` or does it break the constraint machinery somehow?

**Review by isovector — APPROVED** [review]
I love this --- it's so much prettier than the original stuff. Great initiative! Left some stylistic points and some questions, but happy to merge when you're ready.

**Comment by isovector** [comment]
As for why the extra commits, I'm not sure. Do you have an upstream remote set on your local repo? I think it might be using the wrong merge strategy, but it's not a concern --- I squash the commits before merging them anyway.

File: src/Polysemy/ConstraintAbsorber.hs
Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber.hs]: Done.  Sorry, I usually check for unnec. pragmas and forgot.  Do we need the plugin bit in each file?

Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber.hs]: Done.  Sorry, I usually check for unnec. pragmas and forgot.  Do we need the plugin bit in each file?

File: src/Polysemy/ConstraintAbsorber/MonadError.hs
Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber/MonadError.hs]: It might be.  But I hate it.  I could be convinced.  And it doesn't really matter since we aren't exporting it.  ```Orphan``` speaks to part of it's function but not all of it.  And ```Action``` speaks to a different part.  Both seem better than ```Lift``` to me.  ```OrphanAction```?

File: src/Polysemy/ConstraintAbsorber/MonadError.hs
Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber/MonadError.hs]: We ```unsafeCoerce``` it in ```absorbUsingSem```

Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber/MonadError.hs]: That being said, my sense of how this all works is fairly tenuous.  So I hope it makes sense to you!

File: src/Polysemy/ConstraintAbsorber/MonadWriter.hs
Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber/MonadWriter.hs]: We could, I think.  But I'm worried there will come a time when someone needs to do something else there, like a more complex entailment or something and not having access to that argument will be annoying.  But I don't know since my brain frizzes at ```Sub Dict`` ` in the first place.  Should I try to have two versions of ```absorbWithSem```, one simplified and one where you can access that argument just in case?

Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber/MonadWriter.hs]: Actually, right now I can't make that work anyway.  Once I check in the rest of these fixes and you take a look, we can decide if it's worth chasing.

File: src/Polysemy/ConstraintAbsorber.hs
Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber.hs]: done

File: src/Polysemy/ConstraintAbsorber.hs
Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber.hs]: Done.  Sorry, I usually check for unnec. pragmas and forgot.  Do we need the plugin bit in each file?

Comment (isovector) [reply src/Polysemy/ConstraintAbsorber.hs]: The plugin is enabled globally across all of the zoo, so no need for it anywhere.

File: src/Polysemy/ConstraintAbsorber/MonadError.hs
Comment (isovector) [reply src/Polysemy/ConstraintAbsorber/MonadError.hs]: I'm not going to fight you on this! I just name newtypes `Orphan` whenever they're necessary to avoid orphan instances and I'm not going to export them anyway!

File: src/Polysemy/ConstraintAbsorber.hs
Comment (isovector) [inline src/Polysemy/ConstraintAbsorber.hs]: Missing a space here before the final `Type`

File: src/Polysemy/ConstraintAbsorber/MonadError.hs
Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber/MonadError.hs]: Let's leave it then.  Especially since we can always change it without affecting anything else at all.

File: src/Polysemy/ConstraintAbsorber.hs
Comment (isovector) [reply src/Polysemy/ConstraintAbsorber.hs]: I'm also not sure what this kind corresponds to. I'd expect it to be `(Type -> Type) -> Type -> Type`. Where does the extra `Type` come from? Is it the `s` in `MonadState s`?

Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber.hs]: So true.  Fixed.

Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber.hs]: It's the ```s``` you need for Reflection.  The phantom type.

**Comment by adamConnerSax** [comment]
Also, while we're changing so much, should these things be
```absorbReader``` or ```absorbMonadReader``` ?  It's clear enough here but not all monadic constraints are named ```MonadXXX``, e.g., ```PandocMonad``` or ```Resource``` or whatever.

**Comment by isovector** [comment]
I have no strong opinions on the naming. I slightly like `absorbReader` better due to the brevity.

File: src/Polysemy/ConstraintAbsorber/MonadWriter.hs
Comment (isovector) [reply src/Polysemy/ConstraintAbsorber/MonadWriter.hs]: Sounds good. I'm not convinced it's possible --- `Dict`s are weird --- but it's a noble goal if tenable.

Comment (adamConnerSax) [reply src/Polysemy/ConstraintAbsorber/MonadWriter.hs]: "Dicts are weird" seems like a good first album name for a very very geeky punk band.

**Comment by isovector** [comment]
Does this fix #24 when it lands?

**Comment by adamConnerSax** [comment]
It doesn't fix it because it wasn't broken.  I'll close that one since I found my way through.  If we can think of a couple of more standard (than PandocMonad) examples with superclasses or other non-trivial complications, it might be worth putting them in the zoo just so there are more examples for people to look at.  ```Reflection``` and ```Constraints``` are both confusing so using them together is...a lot.

**Comment by isovector** [comment]
I'm hesitant about `PandocMonad` due to #16 :)

**Comment by isovector** [comment]
I'd like to document the fact that this technology exists in `polysemy`. Do you mind if I write a blog post on it, or do you feel like it would be stealing your thunder? Of course, I'll sing your praise throughout such a post!

**Comment by adamConnerSax** [comment]
Yeah.  I was not suggesting ```PandocMonad``` be allowed anywhere near the zoo!  Just that we keep an eye out for other examples that might provide inspiration.  
```PandocMonad``` was trickier, by a bit, than the reflection example, since the functionality in the PandocMonad part of the dictionary did not cover the MonadError stuff so the dictionary had to be expanded, etc. 
So if we can find a low-dependency-weight example that seems useful in the zoo, it might be worth adding just for educational reasons.

**Comment by adamConnerSax** [comment]
Feel free to blog!  You figured out much of it anyway.  I think I just made you do it...
Someday, I'd like to write about these things as well as you do but that day is not now.
I was hoping you'd write a post about it since it will get people using it and maybe building more absorbers.

**Comment by adamConnerSax** [comment]
Also, on an unrelated note, I never remember what "cargo-cult" means.  So I l[ooked it up](https://en.wikipedia.org/wiki/Cargo_cult) and thought "that makes perfect sense!" and then "and that might be offensive to some since the origin of the term is colonialist condescension to native peoples." and, without much looking, that turns out to be [exactly how some people feel](https://twitter.com/justkelly_ok/status/969641245499801600?lang=en).  So I think I'll not use it--someone suggested "Religion without faith" instead, which works but is less catchy. 
Thought it was worth pointing out since I had no idea!

**Comment by adamConnerSax** [comment]
Not sure if you're waiting for my okay on this, but in case you are, it's good to go from my end.  And I have an updated version of ```polysemy-RandomFu``` ready to go once this is on hackage.

**Comment by isovector** [comment]
Fantastic, thanks!

**Review by 1atAlcone — APPROVED** [review]

**Pull request merged**
